pkg/codegen/hcl2/rewrite_convert.go

```diff
@@ -100,37 +97,52 @@ func RewriteConversions(x model.Expression, to model.Type) model.Expression {
 			valueType, diags := to.Traverse(traverser)
 			contract.Ignore(diags)
 
-			item.Key = RewriteConversions(item.Key, model.InputType(model.StringType))
-			item.Value = RewriteConversions(item.Value, model.InputType(valueType.(model.Type)))
+			var valueChanged bool
+			item.Key, _ = rewriteConversions(item.Key, model.InputType(model.StringType))
+			item.Value, valueChanged = rewriteConversions(item.Value, valueType.(model.Type))
+			typecheck = typecheck || valueChanged
 		}
 	case *model.TupleConsExpression:
 		for i := range x.Expressions {
 			valueType, diags := to.Traverse(hcl.TraverseIndex{Key: cty.NumberIntVal(int64(i))})
 			contract.Ignore(diags)
 
-			x.Expressions[i] = RewriteConversions(x.Expressions[i], valueType.(model.Type))
+			var exprChanged bool
+			x.Expressions[i], exprChanged = rewriteConversions(x.Expressions[i], valueType.(model.Type))
+			typecheck = typecheck || exprChanged
 		}
 	case *model.UnaryOpExpression:
-		x.Operand = RewriteConversions(x.Operand, x.OperandType())
+		x.Operand, _ = rewriteConversions(x.Operand, model.InputType(x.OperandType()))
 	}
 
-	// If the expression's type is directly assignable to the destination type, no conversion is necessary.
-	if to.AssignableFrom(x.Type()) && sameSchemaTypes(to, x.Type()) {
-		return x
+	if typecheck {
```

Comment:
I guess I still need to learn how type checking works. If an example is broken, and some bogus value is assigned to a property (e.g. a string to a boolean - before we started converting them, or a non-existing value because of the bug in HCL), why do we still generate those examples? Or what are the cases it catches?

Reply:
> If an example is broken, and some bogus value is assigned to a property (e.g. a string to a boolean - before we started converting them, or a non-existing value because of the bug in HCL), why do we still generate those examples?

In general, we don't convert examples that don't typecheck: e.g. we would not accept an example that passes a list where a string is expected. That said, let's break down the two situations you mentioned, as the second is an exception to the rule.

In the first case, the values aren't actually bogus: [HCL2 performs automatic conversion of these values at runtime](https://github.com/hashicorp/hcl/blob/hcl2/spec.md#type-conversions-and-unification). The rewrite you added matches that behavior; we'll eventually need similar codegen for dynamic conversions of non-literal values.

In the second case, we tolerate missing variables because in some cases they are intended: parts of an example that are not relevant to the resource at hand are often elided. In these cases, we treat the missing variables as having the type `dynamic`, which allows them to pass typechecking.